pkg/scalers/kafka_x_scaler.go

```diff
@@ -0,0 +1,686 @@
+	case KafkaSASLTypeMskIam:
+		_, config := getAwsConfig(metadata.awsRegion,
+			metadata.awsEndpoint,
+			metadata.awsAuthorization)
+
+		saslMechanism = &aws_msk_iam.Mechanism{
+			Signer: sigv4.NewSigner(config.Credentials),
+			Region: metadata.awsRegion,
+		}
+	default:
+		return nil, fmt.Errorf("err sasl type %q given", metadata.saslType)
+	}
+
+	transport := &kafka.Transport{
+		TLS:  tlsConfig,
+		SASL: saslMechanism,
+	}
+	client := kafka.Client{
+		Addr:      kafka.TCP(metadata.bootstrapServers...),
+		Transport: transport,
+	}
+	if err != nil {
+		return nil, fmt.Errorf("error creating kafka client: %w", err)
+	}
+
+	return &client, nil
+}
+
+func (s *kafkaXScaler) getTopicPartitions() (map[string][]int, error) {
+	metadata, err := s.client.Metadata(context.Background(), &kafka.MetadataRequest{
+		Addr: s.client.Addr,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error getting metadata: %w", err)
+	}
+	s.logger.V(4).Info(fmt.Sprintf("Listed topics %v", metadata.Topics))
+
+	if len(s.metadata.topic) == 0 {
+		// in case of empty topic name, we will get all topics that the consumer group is subscribed to
+		describeGrpReq := &kafka.DescribeGroupsRequest{
+			Addr: s.client.Addr,
+			GroupIDs: []string{
+				s.metadata.group,
+			},
+		}
+		describeGrp, err := s.client.DescribeGroups(context.Background(), describeGrpReq)
+		if err != nil {
+			return nil, fmt.Errorf("error describing group: %w", err)
+		}
+		if len(describeGrp.Groups[0].Members) == 0 {
+			return nil, fmt.Errorf("no active members in group %s, group-state is %s", s.metadata.group, describeGrp.Groups[0].GroupState)
+		}
+		s.logger.V(4).Info(fmt.Sprintf("Described group %s with response %v", s.metadata.group, describeGrp))
+
+		result := make(map[string][]int)
+		for _, topic := range metadata.Topics {
+			partitions := make([]int, 0)
+			for _, partition := range topic.Partitions {
+				// if no partitions limitatitions are specified, all partitions are considered
+				if (len(s.metadata.partitionLimitation) == 0) ||
+					(len(s.metadata.partitionLimitation) > 0 && kedautil.Contains(s.metadata.partitionLimitation, int32(partition.ID))) {
+					partitions = append(partitions, partition.ID)
+				}
+			}
+			result[topic.Name] = partitions
+		}
+		return result, nil
+	}
+	result := make(map[string][]int)
+	for _, topic := range metadata.Topics {
+		partitions := make([]int, 0)
+		if kedautil.Contains(s.metadata.topic, topic.Name) {
+			for _, partition := range topic.Partitions {
+				if (len(s.metadata.partitionLimitation) == 0) ||
+					(len(s.metadata.partitionLimitation) > 0 && kedautil.Contains(s.metadata.partitionLimitation, int32(partition.ID))) {
+					partitions = append(partitions, partition.ID)
+				}
+			}
+		}
+		result[topic.Name] = partitions
+	}
+	return result, nil
+}
+
+func (s *kafkaXScaler) getConsumerOffsets(topicPartitions map[string][]int) (map[string]map[int]int64, error) {
+	response, err := s.client.OffsetFetch(
+		context.Background(),
```

Comment:
We should propagate the context until here instead of using a new context

Reply:
done